File: client/web/src/app/core/services/monitoring/sentry-scrubber.ts
import type { Breadcrumb, Event } from '@sentry/angular';

const SESSION_CODE_IN_PATH = /(\/(?:private|ws)\/)[A-Za-z0-9]+/g;

/**
 * Redacts private session codes from URLs/messages
 */
export function scrubSessionCodes(text: string): string {
  return text.replace(SESSION_CODE_IN_PATH, '$1[code]');
}

/**
 * Scrubs one Sentry breadcrumb in place
 */
export function scrubBreadcrumb(breadcrumb: Breadcrumb): Breadcrumb {
  if (breadcrumb.message) {
    breadcrumb.message = scrubSessionCodes(breadcrumb.message);
  }

  for (const key of ['url', 'from', 'to'] as const) {
    const value = breadcrumb.data?.[key];
    if (typeof value === 'string' && breadcrumb.data) {
      breadcrumb.data[key] = scrubSessionCodes(value);
    }
  }
  return breadcrumb;
}

/** Redacts session codes from every string value of span or trace data. */
function scrubData(data: Record<string, unknown> | undefined): void {
  if (!data) return;
  for (const [key, value] of Object.entries(data)) {
    if (typeof value === 'string') {
      data[key] = scrubSessionCodes(value);
    }
  }
}

/**
 * Strips identifying request details and session codes from an error or transaction event in place
 */
export function scrubEvent<T extends Event>(event: T): T {
  event.user = { ip_address: '127.0.0.1' };
  delete event.server_name;
  if (event.request) {
    delete event.request.cookies;
    delete event.request.data;
    delete event.request.query_string;
    if (event.request.url) {
      event.request.url = scrubSessionCodes(event.request.url);
    }
    // The referring page can be a private session URL
    for (const name of Object.keys(event.request.headers ?? {})) {
      if (name.toLowerCase() === 'referer') {
        delete event.request.headers?.[name];
      }
    }
  }
  if (event.transaction) {
    event.transaction = scrubSessionCodes(event.transaction);
  }
  if (event.message) {
    event.message = scrubSessionCodes(event.message);
  }
  for (const exception of event.exception?.values ?? []) {
    if (exception.value) {
      exception.value = scrubSessionCodes(exception.value);
    }
  }
  scrubData(event.contexts?.trace?.data);
  for (const span of event.spans ?? []) {
    if (span.description) {
      span.description = scrubSessionCodes(span.description);
    }
    scrubData(span.data);
  }
  // Browser-derived locale signals
  if (event.contexts?.['device']) {
    delete event.contexts['device']['timezone'];
    delete event.contexts['device']['locale'];
  }
  if (event.contexts?.['culture']) {
    delete event.contexts['culture'];
  }
  return event;
}

File: client/ios/PastePoint/Core/Services/Monitoring/SentryScrubber.swift
//
//  Copyright © 2026 PastePoint. All rights reserved.
//  SPDX-License-Identifier: GPL-3.0-only
//

import Foundation
import Sentry

enum SentryScrubber {

  nonisolated static func anonymousUser() -> User {
    let user = User()
    user.ipAddress = "127.0.0.1"
    return user
  }

  /// Replaces the code in `/private/<code>` and `/ws/<code>` paths with `[code]`.
  nonisolated static func scrubSessionCodes(_ text: String) -> String {
    text.replacing(#/(/(?:private|ws)/)[A-Za-z0-9]+/#) { "\($0.output.1)[code]" }
  }

  /// Scrubs a breadcrumb's message and URL-like data fields.
  nonisolated static func scrub(_ breadcrumb: Breadcrumb) -> Breadcrumb {
    if let message = breadcrumb.message {
      breadcrumb.message = scrubSessionCodes(message)
    }
    for key in ["url", "from", "to"] {
      if let value = breadcrumb.data?[key] as? String {
        breadcrumb.setData(value: scrubSessionCodes(value), key: key)
      }
    }
    return breadcrumb
  }

  /// Strips identity, request details and locale from an event; transactions pass through here too.
  nonisolated static func scrub(_ event: Event) -> Event {
    event.user = anonymousUser()
    event.serverName = nil

    if let request = event.request {
      request.cookies = nil
      request.headers = nil
      request.queryString = nil
      if let url = request.url {
        request.url = scrubSessionCodes(url)
      }
    }

    if var context = event.context {
      context["device"]?.removeValue(forKey: "timezone")
      context["device"]?.removeValue(forKey: "locale")
      context.removeValue(forKey: "culture")
      context["app"]?.removeValue(forKey: "device_app_hash")
      event.context = context
    }
    return event
  }

  /// Drops the user attributes the SDK stamps on every log (its installation id).
  nonisolated static func scrub(_ entry: SentryLog) -> SentryLog {
    entry.body = scrubSessionCodes(entry.body)
    for key in ["user.id", "user.name", "user.email"] {
      entry.attributes.removeValue(forKey: key)
    }
    return entry
  }
}
